feat(bigtable): Managed Backups wrappers

mf2199 opened the pull request:
A draft/beta implementation of Managed Backups' wrappers for Google Cloud Bigtable.

mf2199 commented on google/cloud/bigtable/backup.py:
```diff
@@ -0,0 +1,360 @@
+# Copyright 2015 Google LLC
```
:+1: Indeed, thanks!)

mf2199 commented on google/cloud/bigtable/backup.py:
```diff
@@ -0,0 +1,360 @@
+    r"clusters/(?P<cluster_id>[a-z][-a-z0-9]*)/"
+    r"backups/(?P<backup_id>[a-z][a-z0-9_\-]*[a-z0-9])$"
+)
+_DEFAULT_EXPIRE_PERIOD = 604800  # 7 days
```
Right, this is a leftover from a helper method mentioned in a `TODO` below. The idea was to set a default expiration time as a convenience to the end user, which I later decided to remove. Obviously, this line is now redundant, so thanks for catching it.

mf2199 commented on google/cloud/bigtable/backup.py:
```diff
@@ -0,0 +1,360 @@
+        except NotFound:
+            return None
+
+    def refresh(self):
```
This is one of the [optional] "convenience" methods mentioned during last meeting. The idea is to be able to handle an existing Backup without having to explicitly instantiate a second Backup object. Just thought I'd leave it in for the Draft PR version, although it's a non-essential feature that can be safely removed.

kolea2 replied:
yes, to be consistent with the other libraries I'd prefer leaving it out for now. Thanks!

mf2199 replied:
Done.

mf2199 commented on google/cloud/bigtable/backup.py:
```diff
@@ -0,0 +1,360 @@
+        api.update_backup(backup_update, update_mask)
+        self._expire_time = new_expire_time
+
+    def is_ready(self):
```
If you mean the getter of the `state` property, then that could surely do the job, while this one is another "convenience", making the later code somewhat shorter and hopefully more intuitive. Shall we remove this anyway?

kolea2 replied:
I think state will be fine so we are in line with the other libraries.

mf2199 replied:
Done.

mf2199 commented on google/cloud/bigtable/instance.py:
```diff
@@ -641,6 +644,184 @@ def list_tables(self):
+            self.name + "/clusters/" + cluster_id,
+            filter_=filter_,
+            order_by=order_by,
+            page_size=page_size,
```
According to `bigtable_table_admin.pb2.py`,
```
      page_size:
          Number of backups to be returned in the response. If 0 or
          less, defaults to the server’s maximum allowed page size.
```
In the `list_backups` method of `bigtable_table_admin_client.py`, this parameter defaults to `None`, so here we follow the same logic. But I agree, it should probably default to `0` rather than none-type, will change.

mf2199 commented on google/cloud/bigtable/instance.py:
```diff
@@ -641,6 +644,184 @@ def list_tables(self):
 
         return result
 
+    def backup(self, backup_id, cluster_id=None, table=None, expire_time=None):
```
Agreed, and the same seems to apply for `list_backups` and `restore_table` as well. For now I've moved all three methods, but please correct me if I'm wrong.

mf2199 commented on google/cloud/bigtable/backup.py:
```diff
@@ -0,0 +1,360 @@
+# from google.cloud.bigtable_admin_v2.proto import bigtable_table_admin_pb2
+from google.cloud.exceptions import NotFound
+
+_BACKUP_NAME_RE = re.compile(
```
`re` is a part of Python standard library. Since it is used only once, I moved it into the `from_pb` method, to make testing more explicit. The coverage shows 100% either way.

mf2199 commented:
@kolea2  Added the integration test. To avoid repetitive LROs, everything is placed into a single comprehensive test method inside `TestTableAdminAPI` class. But please let me know if you'd rather see several smaller tests covering each individual Backup function separately.

mf2199 commented:
@kolea2 PTAL. Replaced some string composition code with 'canonical' methods from `bigtable_table_admin_client`; added `__eq__` and `__ne__` convenience methods commonly found in other classes (e.g. `Table`); as well as some other refactoring. Also, giving another thought to the `reload` method that was removed in compliance with other Backups libraries, maybe we ought to have it back, as this is another quite common method, at least in Python APIs. Just a thought.

kolea2 commented:
> Also, giving another thought to the `reload` method that was removed in compliance with other Backups libraries, maybe we ought to have it back, as this is another quite common method, at least in Python APIs. Just a thought.

SGTM for the `reload` method

crwilcox commented on google/cloud/bigtable/backup.py:
```diff
@@ -0,0 +1,392 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""A user-friendly container for Cloud Bigtable Backup."""
```
not sure container is the correct word here? wrapper?. Also `a Cloud Bigtable Backup` or `Cloud Bigtable Backups` perhaps?  

And it seems we mix `google cloud bigtable` and `cloud bigtable`. Maybe best to be consistent?

mf2199 replied:
Agreed about the `wrapper` vs `container` as well as `Google Cloud Bigtable`. Not sure about the article though, since it's a definition of notion here, rather than "a" (i.e. particular) Backup class. I'll put it in for now, just must say that it's debatable.

crwilcox commented on google/cloud/bigtable/backup.py:
```diff
@@ -0,0 +1,392 @@
+    :type backup_id: str
+    :param backup_id: The ID of the backup.
+
+    :type instance: :class:`~google.cloud.spanner_v1.instance.Instance`
```
spanner?  Maybe do a quick check through code for this. My guess is some of this was lifted from the spanner impl?

mf2199 replied:
Totally. Thanks!

crwilcox commented on google/cloud/bigtable/backup.py:
```diff
@@ -0,0 +1,392 @@
+        backup_id = match.group("backup_id")
+        cluster_id = match.group("cluster_id")
+
+        match = re.compile(
```
There is a const regex above. Should this be done similarly?

mf2199 replied:
Makes sense. We couldn't decide whether to declare `re` as constants, hence the two ended up in different representations) It's fixed now.

crwilcox commented on google/cloud/bigtable/backup.py:
```diff
@@ -0,0 +1,392 @@
+        """
+        if not self._expire_time:
+            raise ValueError('"expire_time" parameter must be set')
+            # TODO: Consider implementing a method that sets a default value of
```
Can we create an issue to track this as a feature request?

mf2199 replied:
Certainly, that's the idea. I'd just wait until the initial release is merged, but please let know if you'd rather see it sooner.

crwilcox commented on google/cloud/bigtable/backup.py:
```diff
@@ -0,0 +1,392 @@
+        backup_update = table_pb2.Backup(
+            name=self.name, expire_time=_datetime_to_pb_timestamp(new_expire_time),
+        )
+        # update_mask = {"paths": ["expire_time"]}
```
Can this line be removed, or is it needed?

mf2199 replied:
Absolutely. Sorry, missed it during the cleanup.

crwilcox commented on google/cloud/bigtable/table.py:
```diff
@@ -782,6 +784,180 @@ def mutations_batcher(self, flush_count=FLUSH_COUNT, max_row_bytes=MAX_ROW_BYTES
+        :param backup_id: The ID of the Backup to be created.
+
+        :type cluster_id: str
+        :param cluster_id: (Optional) The ID of the Cluster. Required for
```
Seems odd we have an optional but required parameter. Can we get this by retrieving the backup? This kind of seems required? What is the use case gained by making this optional? Is this just to make it easier to create a client side object?

crwilcox commented on tests/system.py:
```diff
@@ -829,6 +832,58 @@ def test_delete_column_family(self):
+        # Sanity check for `Backup.exists()` method
+        self.assertFalse(temp_backup.exists())
+
+        # Tesing `Backup.create()` method
```
Testing
```suggestion
        # Testing `Backup.create()` method
```

mf2199 replied:
:+1: Thanks!

crwilcox commented on tests/system.py:
```diff
@@ -829,6 +832,58 @@ def test_delete_column_family(self):
+        self.assertFalse(temp_backup.exists())
+
+        # Tesing `Backup.create()` method
+        _ = temp_backup.create().result()
```
I might be missing it, but it doesn't seem we use `_`. What is the value in assigning it then?

mf2199 replied:
Sorry, old habits) Thanks for catching!

crwilcox commented on tests/system.py:
```diff
@@ -829,6 +832,58 @@ def test_delete_column_family(self):
+        # Testing `Table.restore()` and `Backup.retore()` methods
+        restored_table_id = "test-backup-table-restored"
+        restored_table = Config.INSTANCE_DATA.table(restored_table_id)
+        _ = temp_table.restore(
```
Similar to above question.

mf2199 replied:
Same as above)

kolea2 commented on google/cloud/bigtable/table.py:
```diff
@@ -782,6 +784,180 @@ def mutations_batcher(self, flush_count=FLUSH_COUNT, max_row_bytes=MAX_ROW_BYTES
+        )
+        client = self._instance._client.table_admin_client
+        backup_list_pb = client.list_backups(
+            # self._instance.name + "/clusters/" + cluster_id,
```
leftover comment?

mf2199 replied:
Totally. Thanx for pointing out.

kolea2 commented:
@mf2199 lgtm - would you be able to take a look at the build error?

mf2199 commented:
@kolea2 Done. According to the log, one of the errors was due to the `datetime.timestamp` method not existing in Python 2, the support of which is currently being dropped. The Python 3 tests run OK. I've added the backwards compatibility at the expense of somewhat less elegant code.
